src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -227,14 +237,76 @@ protected PathFragment readSymbolicLink(PathFragment path) throws IOException {
+    FileStatus stat = statNullable(linkPath, /* followSymlinks= */ false);
+    if (stat != null) {
+      throw new IOException(linkPath + " (File exists)");
+    }
+
+    // If the output is a regular file or tree artifact, it has "copy" semantics, i.e., its contents
+    // are identical to the target artifact. If the target artifact is stored remotely, then,
+    // instead of creating a symlink in the local filesystem (which would necessarily dangle), we
+    // inject a special type of metadata that is later used by the input prefetcher to download the
+    // original artifact and put the symlink in place.
+    Artifact output = outputArtifactByExecPath.get(linkPath.relativeTo(execRoot));
+    if (output != null) {
+      if (output.isTreeArtifact()) {
+        TreeArtifactValue metadata = getRemoteInputTreeMetadata(targetFragment);
+        if (metadata != null) {
+          // The archived representation is never used on open Source Bazel.
+          checkState(
+              metadata.getArchivedRepresentation().isEmpty(),
+              "symlink to archived tree artifact is not supported");
+          checkState(targetFragment.isAbsolute());
+          injectTree(
+              (SpecialArtifact) output,
+              createTreeArtifactValueForSymlink(
+                  (SpecialArtifact) output, metadata, targetFragment.relativeTo(execRoot)));
+          return;
+        }
+      } else if (!output.isSymlink()) {
+        RemoteFileArtifactValue metadata = getRemoteInputMetadata(targetFragment);
+        if (metadata != null) {
+          checkState(targetFragment.isAbsolute());
+          injectFile(
+              output,
+              createRemoteFileArtifactValueForSymlink(
+                  metadata, targetFragment.relativeTo(execRoot)));
+          return;
+        }
+      }
+    }
+
     super.createSymbolicLink(linkPath, targetFragment);
   }
 
+  private RemoteFileArtifactValue createRemoteFileArtifactValueForSymlink(
+      RemoteFileArtifactValue original, PathFragment targetPath) {
+    return RemoteFileArtifactValue.create(
+        original.getDigest(),
+        original.getSize(),
+        original.getLocationIndex(),
+        original.getActionId(),
+        // Avoid a double indirection when the original artifact is already a symlink.
+        original.getSymlinkTargetExecPath().orElse(targetPath));
+  }
+
+  private TreeArtifactValue createTreeArtifactValueForSymlink(
+      SpecialArtifact output, TreeArtifactValue original, PathFragment targetPath) {
+    TreeArtifactValue.Builder builder = TreeArtifactValue.newBuilder(output);
+    for (Map.Entry<TreeFileArtifact, FileArtifactValue> entry :
+        original.getChildValues().entrySet()) {
+      PathFragment parentRelativePath = entry.getKey().getParentRelativePath();
+      builder.putChild(
+          TreeFileArtifact.createTreeOutput(output, parentRelativePath),
+          createRemoteFileArtifactValueForSymlink(
```

Comment:
IDK but do we want store symlinks for each child file? What could be wrong if we don't do that?

Reply:
I think it's safe to omit them now.